src/SFA.DAS.EmployerFinance.UnitTests/Services/ProviderServiceTests/ProviderServiceCache/WhenIGetAProvider.cs

```diff
@@ -0,0 +1,143 @@
+                .Returns(_testProvider);
+
+            // act
+            var result = await _sut.Get(ukPrn);
+
+            // assert
+            Assert.AreEqual(_testProvider, result);
+        }
+
+        [Test]
+        public async Task AndTheProviderIsInTheCache_ThenTheCacheIsQueriedOnce()
+        {
+            // arrange 
+            long ukPrn = RandomNumber(1, 100);
+
+            _mockInProcessCache
+               .Setup(m => m.Get<EmployerFinance.Models.ApprenticeshipProvider.Provider>(It.IsAny<string>()))
+               .Returns(_testProvider);
+
+            // act
+            var result = await _sut.Get(ukPrn);
+
+            // assert
+            _mockInProcessCache.Verify(m => m.Get<EmployerFinance.Models.ApprenticeshipProvider.Provider>($"{nameof(EmployerFinance.Models.ApprenticeshipProvider.Provider)}_{ukPrn}"), Times.Once);
+            _mockInProcessCache.Verify(m => m.Set(It.IsAny<string>(), It.IsAny<object>(), It.IsAny<TimeSpan>()), Times.Never);
+            _mockProviderService.Verify(m => m.Get(ukPrn), Times.Never);
+        }
+
+        [Test]
+        public async Task AndTheProviderIsNotInTheCache_ThenTheProviderServiceIsCalled()
+        {
+            // arrange
+            long ukPrn = RandomNumber(1, 100);
+
+            _mockInProcessCache
+                .Setup(m => m.Get<EmployerFinance.Models.ApprenticeshipProvider.Provider>(It.IsAny<string>()))
+                .Returns<EmployerFinance.Models.ApprenticeshipProvider.Provider>(null);
+
+            // act
+            var result = await _sut.Get(ukPrn);
+
+            // assert
+            _mockProviderService.Verify(m => m.Get(ukPrn), Times.Once);
+        }
+
+        [Test]
+        public async Task AndTheProviderServiceReturnsAProvider_ThenTheProviderIsStoredInTheCache()
+        {
+            // arrange
+            long ukPrn = RandomNumber(1, 100);
+
+            _mockProviderService
+                .Setup(m => m.Get(ukPrn))
+                .ReturnsAsync(_testProvider);
+
+            // act
+            var result = await _sut.Get(ukPrn);
+
+            // assert
+            _mockInProcessCache.Verify(m => m.Set($"{nameof(EmployerFinance.Models.ApprenticeshipProvider.Provider)}_{ukPrn}", _testProvider, It.IsAny<TimeSpan>()), Times.Once);
+        }
+
+        [Test]
+        public async Task AndTheProviderServiceReturnsANullProvider_ThenTheProviderIsNotStoredInTheCache()
+        {
+            // arrange
+            long ukPrn = RandomNumber(1, 100);
+
+            _mockProviderService
+                .Setup(m => m.Get(ukPrn))
+                .Returns(Task.FromResult<EmployerFinance.Models.ApprenticeshipProvider.Provider>(null));
+
+            // act
+            var result = await _sut.Get(ukPrn);
+
+            // assert
+            _mockInProcessCache.Verify(m => m.Set($"{nameof(EmployerFinance.Models.ApprenticeshipProvider.Provider)}_{ukPrn}", _testProvider), Times.Never);
```

Comment:
Would it be better to make sure that no provider (not just the test one) is stored just in case someone writes logic to create a different one?

Reply:
Changed to accept any provider